src/audio/eq_fir/eq_fir_hifi3.c

```diff
@@ -98,64 +92,62 @@ void eq_fir_2x_s24(struct fir_state_32x16 fir[], struct input_stream_buffer *bso
 	ae_int32 z0;
 	ae_int32 z1;
 	ae_int32 *src = (ae_int32 *)source->r_ptr;
-	ae_int32 *snk = (ae_int32 *)sink->w_ptr;
-	ae_int32 *x;
-	ae_int32 *y;
+	ae_int32 *dst = (ae_int32 *)sink->w_ptr;
+	ae_int32 *in;
+	ae_int32 *out;
 	int ch;
-	int i;
+	int i, n, nmax;
 	int rshift;
 	int lshift;
 	int shift;
 	int nch = source->channels;
 	int inc_nch_s = nch * sizeof(int32_t);
-
-	for (ch = 0; ch < nch; ch++) {
-		/* Get FIR instance and get shifts.
-		 */
-		f = &fir[ch];
-		fir_get_lrshifts(f, &lshift, &rshift);
-		shift = lshift - rshift;
-
-		/* Copy src to x and advance src with dummy load */
-		fir_comp_setup_circular(source);
-		x = src;
-		AE_L32_XC(d0, src, sizeof(int32_t));
-
-		/* Copy snk to y0 and advance snk with dummy load. Pointer
-		 * y1 is set to be ahead of y0 with one frame.
-		 */
-		fir_comp_setup_circular(sink);
-		y = snk;
-		AE_L32_XC(d0, snk, sizeof(int32_t));
-
-		for (i = 0; i < (frames >> 1); i++) {
-			/* Load two input samples via input pointer x */
-			fir_comp_setup_circular(source);
-			AE_L32_XC(d0, x, inc_nch_s);
-			AE_L32_XC(d1, x, inc_nch_s);
-
-			/* Convert Q1.23 to Q1.31 compatible format */
-			d0 = AE_SLAA32(d0, 8);
-			d1 = AE_SLAA32(d1, 8);
-
-			/* Compute FIR */
+	int samples = nch * frames;
+
+	while (samples) {
+		nmax = audio_stream_samples_without_wrap_s24(sink, dst);
+		n = MIN(nmax, samples);
+		nmax = audio_stream_samples_without_wrap_s24(source, src);
+		n = MIN(n, nmax);
+		for (ch = 0; ch < nch; ch++) {
+			/* Get FIR instance and get shifts.*/
+			f = &fir[ch];
+			fir_get_lrshifts(f, &lshift, &rshift);
+			shift = lshift - rshift;
+			/* set f->delay as circular buffer */
 			fir_core_setup_circular(f);
-			fir_32x16_2x_hifi3(f, d0, d1, &z0, &z1, shift);
 
-			/* Shift and round to Q1.23 format */
-			d0 = AE_SRAI32R(z0, 8);
-			d0 = AE_SLAI32S(d0, 8);
-			d0 = AE_SRAI32(d0, 8);
+			in = src + ch;
+			out = dst + ch;
+
+			for (i = 0; i < (n >> 1); i += nch) {
+				/* Load two input samples via input pointer x */
+				AE_L32_XP(d0, in, inc_nch_s);
+				AE_L32_XP(d1, in, inc_nch_s);
```

Comment:
this processing is still based on channel data, i.e, try to handle one channel data in this line 123 for loop, then handle another channel, so:

1. Possible handle data based on sample? take 2 channel as example.
    handle L1R1L2R2
    current L1 L2 --> R1 R2
    proposed: L1R1L2R2 --> L3R3L4R4.
you can still handle 2 samples for each channel, this will make load/store fully aligned with 64 bit operation.

another disadvantage with current handle is: once the buffer is big, take deep buffer for example.
the cache miss will cost much more cycles than real, image big buffer have to write back and load new data in cache.

Reply:
each channel can have its own coefficient, and its delay depends on the statistic in this channel, processing by sample will cost much more cycles.